Constructor-io/constructorio-catalog-ingestor
# [PARTI-202] Catalog Ingestor: Enable ingestions to partially update an item

# PR Description



This PR adds the following changes:

-  Add the `PATCH` ingestion type
- Updates documentation accordingly
- Adjust the `CatalogIngestionPayload` type to reflect the different required catalog fields between `PATCH` vs the other two

## ⚠️ Dependencies



- None

## 🕹 Demo



Performing a `PATCH` ingestion:
<img  src="https://user-images.githubusercontent.com/29166076/213785290-73cf97da-033b-461f-a20a-58b55c10ee6f.png" height="250"/>

| Before | After |
| -- | -- | 
| ![image](https://user-images.githubusercontent.com/29166076/213785338-0ccd6003-0b9a-4b11-94a0-2e3aab110042.png) | ![image](https://user-images.githubusercontent.com/29166076/213785386-67bdfbf6-995e-45c2-8cff-d55fba11aeed.png) |



# How to test



## 🧪 What should be tested?

As a reviewer, you should test:

- Check items being updated
- Check groups being updated
- Check variations being updated
- Make sure only the desired fields are being updated
- Make sure the `ingest`  return type is working as expected
  - when type == `PATCH` => only id is required
  - when other => multiple fields must be required

## 📝 How to test?

Reproduce the following steps to test:

1. Replace the src/index.ts file content with:
```ts
import { CatalogIngestor } from "./catalogIngestor/index";
import { CatalogIngestionType } from "./catalogIngestor/types";

export * from "./catalogIngestor/types";
export * from "./catalogIngestor/index";

const catalogIngestor = new CatalogIngestor({
  apiToken: "<token>",
  apiKey: "<key>",
});

async function test() {
  await catalogIngestor.ingest(async () => {
    return {
      type: CatalogIngestionType.PATCH,
      data: {
        groups: [],
        items: [
          {
            id: "dev-tshirt",
            item_name: "Developer t-shirt - After Patch",
          },
        ],
        variations: [],
      },
    };
  });
}

// eslint-disable-next-line @typescript-eslint/no-floating-promises
test();
```
2. Modify it with your testing data.
3. Check `items` results directly in the customer dashboard
4. Check `groups` and `variations` using the node client, calling the `retreiveVariations` and `getItemGroups` functions
